libminifi/include/utils/ValueUtils.h

```diff
@@ -0,0 +1,203 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenseas/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef NIFI_MINIFI_CPP_VALUEUTILS_H
+#define NIFI_MINIFI_CPP_VALUEUTILS_H
+
+#include <exception>
+#include <string>
+#include <cstring>
+#include <vector>
+#include <cstdlib>
+#include "PropertyErrors.h"
+#include <type_traits>
+#include <limits>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+class ValueParser {
+ private:
+  template< class... >
+  using void_t = void;
+
+  template<typename From, typename To, typename = void>
+  struct is_non_narrowing_convertible: std::false_type {
+    static_assert(std::is_integral<From>::value && std::is_integral<To>::value, "Checks only integral values");
+  };
+
+  template<typename From, typename To>
+  struct is_non_narrowing_convertible<From, To, void_t<decltype(To{std::declval<From>()})>>: std::true_type{
+    static_assert(std::is_integral<From>::value && std::is_integral<To>::value, "Checks only integral values");
+  };
+  
+ public:
+  ValueParser(const std::string& str, std::size_t offset = 0): str(str), offset(offset) {}
+
+  template<typename Out>
+  ValueParser& parseInt(Out& out) {
+    static_assert(is_non_narrowing_convertible<int, Out>::value, "Expected lossless conversion from int");
+    try {
+      char *end;
+      long result{std::strtol(str.c_str() + offset, &end, 10)};
+      offset = end - str.c_str();
+      if (result < (std::numeric_limits<int>::min)() || result > (std::numeric_limits<int>::max)()) {
```

Comment:
Maybe this wouldn't work (due to `-Wnarrow`), but wouldn't this simplify things?
```c++
if(out != static_cast<int>(out))
{
    throw ...;
}
return { result };
```

Reply:
this wouldn't work for two reasons: 
- as you said the return expression prevents narrowing conversions
- casting to signed integrals, if the source value cannot be represented in the target type, is implementation defined (might even raise a signal)